Fix email theme

lchanmann opened the pull request:
## Ref
- https://nanopay.atlassian.net/browse/NP-2368

## Changes
- Add spid property to Theme
- Support finding theme through user.SPID
- Update EmailUtility to try getting theme from the context before going through theme finding

## Cases
1. domain:localhost => load theme by user.SPID, then group and user
2. domain:other => load theme by domain, then group and user
    - when themeDomain not found, fallback to 1. domain:localhost
    - when themeDomain is found but associated theme not found => load theme by user.SPID, then group and user
3. when loading theme by `user.SPID: 'x.y.z'` not found, try `user.SPID: 'x.y'` then `user.SPID: 'x'` then `user.SPID: '*'`

lchanmann committed:
Load theme before performing find

lchanmann committed:
Load current theme into the system context

so that rules that send notfication would use the current theme.

lchanmann committed:
Find theme by the default theme domain for spid

lchanmann committed:
Use user spid as domain

lchanmann committed:
Add spid property to theme domain and find theme support

lchanmann commented on src/foam/nanos/ruler/RuleEngine.java:
```diff
@@ -29,7 +29,7 @@
 
   public RuleEngine(X x, X systemX, DAO delegate) {
-    setX(systemX);
+    setX(systemX.put("theme", x.get("theme")));
```
Carry the current theme to the system context so that we won't need to change every rules that need the system context and will also trigger notification.

kgrgreer replied:
If the RuleEngine is first accessed from something like a CRON Job, then 'x' will be the systemX and it won't contain a theme. If it's accessed from the web, then it will just be stuck with the theme of the first user to access this service, which will be difficult to debug in the future.

jlhughes replied:
yes, if manually run a cron/script, then you'll get that users (admin) theme, which is not what we want for the rule run.

jlhughes replied:
An approach discussed earlier:  add spid to Theme and only support one Theme per spid.  The ThemeDomains can continue to handle the many domains per whilelabel - all pointing to the same Theme - mind you, no way to enforce this.

lchanmann replied:
RuleEngine is a new instance per execution, the subsequent won't be affected by the former.

kgrgreer replied:
SPID's are meant to be hierarchical, so a company could have as many as they need if they want multiple themes.

jlhughes replied:
@kgrgreer I'm good with this approach.

lchanmann replied:
SPID is added to theme domain, this approach expects the users of a specific SPID to use/come from the same domain.

jlhughes replied:
Our current ServiceProvider implementation does not support hierarchy.

jlhughes replied:
If we just support one theme per spid then spid goes on Theme model.  We can still support Theme per group and this is independent of spid supporting hierarchy. 

web -> ThemeDomain -> theme
cron -> user (spid) [-> group] -> theme

lchanmann committed:
Merge branch 'master' into NP-2368/FixEmailTheme

lchanmann committed:
Set subject before sending notification

lchanmann committed:
Add spid to Theme instead of ThemeDomain

For finding theme via user.SPID, it'd only load the first theme it found until we support hierarchical SPID.
As a consequence, nanopay theme will be loaded for nanopay/ablii users instead of the ablii theme.

lchanmann committed:
Support loading theme via hierarchical SPID

lchanmann committed:
Handle loading theme via user.SPID when cannot acquire via domain or when it uses/regresses to localhost

lchanmann committed:
Merge branch 'master' into NP-2368/FixEmailTheme

lchanmann committed:
Cleanup

lchanmann committed:
Add enabled:true predicate

jlhughes commented on src/foam/nanos/theme/Themes.js:
```diff
@@ -96,31 +98,74 @@ Later themes:
+        var spid = user.getSpid();
+        while ( ! SafetyUtil.isEmpty(spid) ) {
+          theme = (Theme) themeDAO.find(
+            MLang.AND(
```
So we are looking a Theme with id that starts with the spid? 
Why not just by spid - Theme.SPID

lchanmann replied:
yeah, just realized and updated.

lchanmann committed:
Use EQ instead

lchanmann committed:
Load theme via user.SPID on the client side
